### PR TITLE
Allows to change the grid size.

### PR DESCRIPTION
**Pull request by vnmabus**
The code is based in the modifications introduced in https://github.com/GAA-UAM/scikit-fda/commit/b74a84c4f8b59dfba7d78519c5093fd3172b1092#diff-37f5178dc37f3176a7256bdbd91847012d4fe55d6cc523a08e506a78c2e60f7d.

The idea is to discuss the necessary changes, as in that commit most files were not used.

Things remaining:
- [x] Expose the size parameter in Cython.
- [x] Test with Valgrind.

### CODE REVIEW

**Comment by jdtuck** [comment]
The changes to `dp_grid.c` look good. I see no problem in modifying the main routine in the c file and accessing it through the python interface. Would be good to have solid documentation on what it means and a default value assigned. Which you pretty much have in the files from the commit.

File: src/DynamicProgrammingQ2.c [hunk 1]
```diff
@@ -2,30 +2,38 @@
...
 
 void DynamicProgrammingQ2(double *Q1, double *T1, double *Q2, double *T2, int m1, int n1, int n2,
 double *tv1, double *tv2, int n1v, int n2v, double *G, double *T, double *size, double lam1){
   int *idxv1 = 0;
   int *idxv2 = 0;
   double *E = 0; /* E[ntv1*j+i] = cost of best path to (tv1[i],tv2[j]) */
   int *P = 0; /* P[ntv1*j+i] = predecessor of (tv1[i],tv2[j]) along best path */
+  size_t nbhd_count; /* Number of indexes */
+
+  size_t nbhd_dim = 7;
```
Comment (jdtuck) [inline src/DynamicProgrammingQ2.c:15]: assuming you are hardcoding for testing right now

Comment (vnmabus) [reply src/DynamicProgrammingQ2.c:15]: Yes, I was trying to do it one step at a time. I have to convert it to a parameter and pass it everywhere, but I think that I will do that tomorrow 😅.

Comment (jdtuck) [reply src/DynamicProgrammingQ2.c:15]: No problem, thanks again.

File: fdasrsf/utility_functions.py [hunk 2]
```diff
@@ -131,6 +131,7 @@ def optimum_reparam(q1, time, q2, method="DP", lam=0.0, f1o=0.0, f2o=0.0):
     :param q2: vector of size N or array of NxM samples samples of second SRSF
     :param method: method to apply optimization (default="DP") options are "DP" and "DP2"
```
Comment (jdtuck) [inline fdasrsf/utility_functions.py:132]: Lets make the default DP2 now

File: fdasrsf/utility_functions.py [hunk 3]
```diff
@@ -122,7 +122,7 @@ def srsf_to_f(q, time, f0=0.0):
     return f
 
 
-def optimum_reparam(q1, time, q2, method="DP", lam=0.0, f1o=0.0, f2o=0.0):
+def optimum_reparam(q1, time, q2, method="DP", lam=0.0, f1o=0.0, f2o=0.0, grid_dim=7):
```
Comment (jdtuck) [inline fdasrsf/utility_functions.py:125]: change default to DP2

File: src/dp_grid.c [hunk 4]
```diff
@@ -4,85 +4,31 @@
...
   int sr, sc;  /* source row and column */
   int tr, tc;  /* target row and column */
   double w, cand_cost;
   int i;
   
   E[0] = 0.0;
-  for ( i=1; i<ntv1; E[i++]=1e6 );
-  for ( i=1; i<ntv2; E[ntv1*i++]=1e6 );
+  for ( i=1; i<ntv1; E[i++]=INFINITY );
+  for ( i=1; i<ntv2; E[ntv1*i++]=INFINITY );
```
Comment (jdtuck) [inline src/dp_grid.c:23]: I should have done this long ago

File: fdasrsf/utility_functions.py (hunk 2 above)
Comment (jdtuck) [reply fdasrsf/utility_functions.py:132]: Good question, they are both the same in the sense of the algorithm, however DP2 can handle different sample spacing for q1 and q2. DP is the original code from the work and kept for comparison. Anuj prefers it as it was the original and I have kept it the default for him.

Comment (vnmabus) [reply fdasrsf/utility_functions.py:132]: But the results were very different with the two methods, at least in some of my tests.

Comment (jdtuck) [reply fdasrsf/utility_functions.py:132]: So I should qualify, they are both implementing Dynamic Programming, they are not the same in implementation. I wrote the second version and have not wrote DP. The differences come from implementation of the spline.

**Comment by vnmabus** [comment]
I should test this in Valgrind for invalid writes and leaks, but I am currently in Windows.

**Comment by jdtuck** [comment]
Ok, that would be good. Know if a good way to add that to Travis to make it easy?

**Comment by jdtuck** [comment]
Ok, I might need to think of way to add valgrind to Travis, but could be a big pain.

**Comment by vnmabus** [comment]
Is this change already in the released version?

**Comment by jdtuck** [comment]
Not yet, I have one more feature to finish today and I was going to release a new version. I have rewritten the RBFGS option and just testing now

**Comment by jdtuck** [comment]
New release on pip with the changes and rbfgs optimization now as well

**Pull request merged**